File: lib/legacy_rails/app/models/media_run.rb
class MediaRun < ApplicationRecord
  belongs_to :media_sequence, optional: true    #to be replaced by a has_many in future releases for reusable media_runs
  belongs_to :media_piece
  belongs_to :marketer
  has_many :bids
  has_many :ad_events

  before_validation :default_values

  validates :banner_retry_buffer_hours, :frequency, :frequency_buffer_hours, :maximum_banner_count, :media_piece_id, :marketer_id,
    presence: true

  validates :banner_retry_buffer_hours, :frequency, :frequency_buffer_hours, :maximum_banner_count,
    numericality: true

  def descriptive_name
    "#{media_piece.display_url} | #{media_piece.title} | #{frequency}:#{frequency_buffer_hours}:#{maximum_banner_count}:#{banner_retry_buffer_hours}"
  end

  private
  def default_values
    self.is_active = true
    self.marketer_id ||= 1
    self.sequence_start_phase ||= 1
    self.sequence_end_phase ||= self.sequence_start_phase
  end
end

File: lib/qlarius/sponster/campaigns/media_run.ex
defmodule Qlarius.Sponster.Campaigns.MediaRun do
  use Ecto.Schema
  import Ecto.Changeset

  alias Qlarius.Sponster.Campaigns.MediaSequence
  alias Qlarius.Sponster.Ads.MediaPiece
  alias Qlarius.Sponster.Marketer

  @primary_key {:id, :id, autogenerate: true}
  @timestamps_opts [type: :naive_datetime, inserted_at: :created_at, updated_at: :updated_at]

  schema "media_runs" do
    field :sequence_start_phase, :integer
    field :sequence_end_phase, :integer
    field :frequency, :integer
    field :frequency_buffer_hours, :integer
    field :maximum_banner_count, :integer
    field :banner_retry_buffer_hours, :integer
    field :is_active, :boolean

    belongs_to :marketer, Marketer
    belongs_to :media_piece, MediaPiece
    belongs_to :media_sequence, MediaSequence

    has_many :offers, Qlarius.Sponster.Offer
    has_many :ad_events, Qlarius.Sponster.AdEvent

    timestamps()
  end

  def changeset(media_run, attrs) do
    media_run
    |> cast(attrs, [
      :sequence_start_phase,
      :sequence_end_phase,
      :frequency,
      :frequency_buffer_hours,
      :maximum_banner_count,
      :banner_retry_buffer_hours,
      :is_active,
      :marketer_id,
      :media_piece_id,
      :media_sequence_id
    ])
    |> validate_required([
      :media_piece_id,
      :media_sequence_id,
      :marketer_id
    ])
    |> foreign_key_constraint(:marketer_id)
    |> foreign_key_constraint(:media_piece_id)
    |> foreign_key_constraint(:media_sequence_id)
  end
end
